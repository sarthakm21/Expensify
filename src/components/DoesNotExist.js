import React from "react";
const DoesNotExist = () => <div>404! The page requested does not exist.</div>;
export default DoesNotExist;
