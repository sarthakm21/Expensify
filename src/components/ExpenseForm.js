import React, { Component } from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends Component {
  //Using state directly without constructor because of the babel plugin "@babel/plugin-proposal-class-properties",
  //Using class properties instead of methods for the same reason
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount.toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: "",
    };
  }

  //Declaring them as class properties instead of methods(by using arrow function) to prevent using 'this' binding
  /* Never use e.target.value directly in the setState because its callback is asynchronous and by the time
    it executes, it is possible that the value no longer exists! */
  handleDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  handleNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  handleAmountChange = (e) => {
    const amount = e.target.value;
    (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) &&
      this.setState(() => ({ amount }));
    //Using a regex to match favourable strings. Refer http://www.regex101.com
    console.log(this.state.amount);
  };

  handleSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please fill out both description and amount",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  // -------------------SingleDatePicker Component's event handlers ------------------------------------

  handleDateChange = (createdAt) => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  handleFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  render() {
    return (
      <div>
        {this.state.error}
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.handleDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.handleDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.handleFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false} //To allow user to enter the dates in the past as well
            displayFormat="DD MMM YYYY"
          />
          <textarea
            placeholder="Add some note for the expense"
            value={this.state.note}
            onChange={this.handleNoteChange}
          ></textarea>
          <button>{this.props.text}</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
