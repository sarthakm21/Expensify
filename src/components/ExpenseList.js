import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
  <div>
    {props.expenses.length === 0
      ? "Add an expense"
      : props.expenses.map((expense) => (
          <ExpenseListItem key={expense.id} {...expense} />
          //Using spread syntax passes props with the same key value pair as that present in the expense object
        ))}
  </div>
);

//Return value of connect is a HOC to which we pass the component in which we want to use redux store
//The connect function takes in a callback function that lets us pass the values of the store that will be passed to
//our Wrapped component as props. Here we only need to pass the expenses array.
const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
