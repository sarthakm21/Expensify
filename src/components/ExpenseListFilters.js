import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setFilterText,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }));
  };

  onTextChange = (e) => {
    this.props.setFilterText(e.target.value);
  };

  onSortChange = (e) => {
    e.target.value === "date"
      ? this.props.sortByDate()
      : this.props.sortByAmount();
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="start"
          endDate={this.props.filters.endDate}
          endDateId="end"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          displayFormat="DD MMM YYYY"
        />
      </div>
    );
  }
}

const mapState = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatch = (dispatch) => ({
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setFilterText: (text) => dispatch(setFilterText(text)),
  setStartDate: (date) => dispatch(setStartDate(date)),
  setEndDate: (date) => dispatch(setEndDate(date)),
});
export default connect(mapState, mapDispatch)(ExpenseListFilters);
