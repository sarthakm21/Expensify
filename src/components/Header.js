import React from "react";
import { NavLink } from "react-router-dom";

//We use Link to render a page from Client Side directly and do not need the App to request the server first, reducing loading time
//NavLink is similar to link, with extra props like activeClassName and some more. Check the docs for more
const Header = () => (
  <div>
    <h1>Expensify</h1>
    <NavLink to="/create" activeClassName="isActive">
      Create Expense
    </NavLink>
    <NavLink to="/help" activeClassName="isActive">
      Help
    </NavLink>
    <NavLink to="/" activeClassName="isActive" exact={true}>
      Dashboard
    </NavLink>
  </div>
);

export default Header;
