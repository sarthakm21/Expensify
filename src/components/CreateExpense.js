import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export class CreateExpense extends React.Component {
  onSubmit = (expense) => {
    console.log(expense)
    this.props.addExpense(expense);
    this.props.history.push("/");
  }

  render() {
    return (
      <div>
        <div>
          This is the create element
          <ExpenseForm
            onSubmit={this.onSubmit}
            text="Add Expense"
          />
        </div>
      </div>
    );
  }
}

//aka mapDispatchToProps. Does the same thing to dispatch that mapStateToProps does to state
const mapDispatch = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatch)(CreateExpense);
