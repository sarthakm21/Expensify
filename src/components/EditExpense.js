import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpense extends React.Component {
  onSubmit = (expense) => {
    //Func. gets called by the submit event handler in the form which passes the new data as a parameter
    console.log(expense);
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id); //Can also use props.match.params.id since both are same
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense} //The expense received from mapStateToProps, i.e. the expense with an id same as the props.match.params.id
          onSubmit={this.onSubmit}
          text="Edit Expense"
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(
    (expense) => expense.id === props.match.params.id
  ),
});

const mapDispatch = (dispatch) => ({
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
});

export default connect(mapStateToProps, mapDispatch)(EditExpense);
