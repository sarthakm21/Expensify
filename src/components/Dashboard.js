import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpenseSummary";

const Dashboard = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default Dashboard;
