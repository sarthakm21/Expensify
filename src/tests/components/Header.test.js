import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { Header } from "../../components/Header";

let wrapper, startLogout;
beforeEach(() => {
  startLogout = jest.fn();
  wrapper = shallow(<Header startLogout={startLogout} />);
});

test("should render Header component", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should call startLogout on button click", () => {
  wrapper.find("button").simulate("click");
  expect(startLogout).toHaveBeenCalled();
});
