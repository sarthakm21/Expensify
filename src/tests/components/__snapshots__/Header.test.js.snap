// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header component 1`] = `
<div>
  <h1>
    Expensify
  </h1>
  <NavLink
    activeClassName="isActive"
    to="/create"
  >
    Create Expense
  </NavLink>
  <NavLink
    activeClassName="isActive"
    to="/help"
  >
    Help
  </NavLink>
  <NavLink
    activeClassName="isActive"
    exact={true}
    to="/"
  >
    Dashboard
  </NavLink>
</div>
`;
